### PR TITLE
fixed a text wrapping issue caused by layout height desync

### PR DESCRIPTION
**Pull request by XiaotianNetlift**
GitHub Issue (If applicable): #

## PR Type
What kind of change does this PR introduce?
- Bugfix

## What is the current behavior?
Currently when the `TextBlock` changes layout to accommodate for the available height, the height measured with the old layout is returned.

## What is the new behavior?
The height is recalculated on layout changes.

## PR Checklist
Please check if your PR fulfills the following requirements:
- [x] Tested code with current [supported SDKs](https://github.com/unoplatform/uno/blob/master/README.md#uno-features)
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] [Wasm UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Updated the [Release Notes](https://github.com/unoplatform/uno/tree/master/doc/ReleaseNotes)
- [x] Associated with an issue (GitHub or internal)

## Other information
Internal Issue (If applicable):
[AB#164480](https://dev.azure.com/nventive/67eedefd-9293-4f70-a624-89ecbe8ad34f/_workitems/edit/164480)
https://github.com/unoplatform/private/issues/83

### CODE REVIEW

**Comment by XiaotianNetlift** [comment]
For context:
```csharp
@ Windows.UI.Xaml.Controls.TextBlock.LayoutBuilder.UpdateLayout
    int linecount = Layout.LineCount; // 2
    var measuredHeight = Layout.GetLineTop(linecount); // 108
    
    if (maxHeight != null && measuredHeight > maxHeight) // true
        var lineAtHeight = Layout.GetLineForOffset(maxHeight.Value) + 1; // 2
        measuredHeight = Layout.GetLineTop(lineAtHeight);  // 108

        // [orignal source comment:] We don't want to display half of a line so we remove it 
        if (linecount > 1 && measuredHeight > maxHeight) // true = 2 > 1 && 108 > 105
            lineAtHeight = Math.Max(0, lineAtHeight - 1); // 1 = max(0, 2 - 1)
        
        // here the layout got replaced with a `BoringLayout`
        MakeLayout(desiredWidth, maxLines: lineAtHeight);
        @ Windows.UI.Xaml.Controls.TextBlock.LayoutBuilder.MakeLayout
            if (maxLines == 1) // true
                _metrics = BoringLayout.IsBoring(_textFormatted, _paint, _metrics);
                if (_metrics != null) // true
                    Layout = 
                        (Layout as BoringLayout)?.ReplaceOrMake(...) ??
                        new BoringLayout(...)

	// but, we are still returning the old cached measured height which was for 2 lines... 
	// this value is already outdated, since we replaced the layout, by capping the maxlines to 1
    return new Size(PhysicalToLogicalPixels(Layout.Width), PhysicalToLogicalPixels(measuredHeight)); // (912, 108) / device_scaling (nve214: 2.625)
```

File: src/Uno.UI/UI/Xaml/Controls/TextBlock/TextBlock.Android.cs [hunk 1]
```diff
@@ -572,6 +572,8 @@ private Size UpdateLayout(Size availableSize, bool exactWidth = false)
 						desiredWidth,
 						maxLines: lineAtHeight
 					);
+
+					measuredHeight = Layout.GetLineTop(Layout.LineCount);
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/TextBlock/TextBlock.Android.cs]: Please add a comment here noting why this is important.

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Controls/TextBlock/TextBlock.Android.cs]: Repro is very important indeed, and also please take a look at possible regression in screenshot tests.

**Comment by XiaotianNetlift** [comment]
@davidjohnoliver @jeromelaban I've added the sample.
pr1954-droid-text-wrapping

**Comment by jeromelaban** [comment]
Thanks for the details! Just make sure the last state of the sample demonstrates the fixed issue, e.g. goes through all the steps needed to recreate the bug, then leaves the control in that state so there`s a screen shot taken there.

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/TextBlockControl/TextBlock_TextWrapping_PR1954_EdgeCase.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+			switch ((sender as Windows.UI.Xaml.Controls.Button)?.Content)
+			{
+				case "+": return true;
+				case "-": return false;
+
+				default: return default;
+			}
+		}
+
+		//private delegate Size UpdateLayoutSignature(ref TextBlock.LayoutBuilder layout, Size availableSize, bool exactWidth)
```
Comment (davidjohnoliver) [inline src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/TextBlockControl/TextBlock_TextWrapping_PR1954_EdgeCase.xaml.cs]: Remove the commented code

**Pull request merged**